### ISSUE TITLE
Issue with spring-data "startingWith" and hibernate 5.6.7: Parameter value [\] did not match expected type [java.lang.String (n/a)]" with findAllByXXXStartingWith

### ISSUE THREAD

**Issue by glhez** [spring-projects/spring-data-jpa#2472]
Hello,

I am trying to fetch some entity using a “find all by property starting with” query which amount to a CriteriaQuery using javax.persistence.criteria.CriteriaBuilder.like(Expression<String>, String, char)

When migrating to hibernate-entitymanager 5.6.7.Final I do have this error:

```
java.lang.IllegalArgumentException: Parameter value [\] did not match expected type [java.lang.String (n/a)]
	at org.hibernate.query.spi.QueryParameterBindingValidator.validate(QueryParameterBindingValidator.java:54) ~[hibernate-core-5.6.7.Final.jar:5.6.7.Final]
	at org.hibernate.query.spi.QueryParameterBindingValidator.validate(QueryParameterBindingValidator.java:27) ~[hibernate-core-5.6.7.Final.jar:5.6.7.Final]
	at org.hibernate.query.internal.QueryParameterBindingImpl.validate(QueryParameterBindingImpl.java:90) ~[hibernate-core-5.6.7.Final.jar:5.6.7.Final]
	at org.hibernate.query.internal.QueryParameterBindingImpl.setBindValue(QueryParameterBindingImpl.java:55) ~[hibernate-core-5.6.7.Final.jar:5.6.7.Final]
	at org.hibernate.query.internal.AbstractProducedQuery.setParameter(AbstractProducedQuery.java:501) ~[hibernate-core-5.6.7.Final.jar:5.6.7.Final]
	at org.hibernate.query.internal.AbstractProducedQuery.setParameter(AbstractProducedQuery.java:122) ~[hibernate-core-5.6.7.Final.jar:5.6.7.Final]
	at org.hibernate.query.criteria.internal.compile.CriteriaCompiler$1$1.bind(CriteriaCompiler.java:141) ~[hibernate-core-5.6.7.Final.jar:5.6.7.Final]
	at org.hibernate.query.criteria.internal.CriteriaQueryImpl$1.buildCompiledQuery(CriteriaQueryImpl.java:364) ~[hibernate-core-5.6.7.Final.jar:5.6.7.Final]
	at org.hibernate.query.criteria.internal.compile.CriteriaCompiler.compile(CriteriaCompiler.java:171) ~[hibernate-core-5.6.7.Final.jar:5.6.7.Final]
	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:774) ~[hibernate-core-5.6.7.Final.jar:5.6.7.Final]
	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:114) ~[hibernate-core-5.6.7.Final.jar:5.6.7.Final]
```

It seems to happens when invoking twice the same method (in the attached [demo_spring-data-2.6.3_hibernate_5.6.7..zip][1], `findByLastNameStartingWith`):

In the first invocation, we can see the validate (`org.hibernate.query.spi.QueryParameterBindingValidator.validate(Type, Object, TemporalType)`) takes the following arguments set:

- null, '\', null
- org.hibernate.type.StringType, ‘Bauer%’, null

In the second case:
- org.hibernate.type.StringType, '\', null => FAILS (StringType expect a String, but got a Character)

The demo (was generated using Spring Initializer and this doc: https://www.baeldung.com/spring-data-jpa-query: [demo_spring-data-2.6.3_hibernate_5.6.7..zip][1]. It can be imported as Maven project in Eclipse and IntelliJ.

**Note:** I did try using `CriteriaQuery` as shown below and got no error and that's why I created issue here instead of hibernate, since I'm not sure where the problem might be (I would say that it is in Hibernate but given the example below works, there might be more to it):

```
    private static void demonstrateWithCriteria(EntityManager em) {
        CriteriaBuilder b = em.getCriteriaBuilder();
        CriteriaQuery<Customer> q = b.createQuery(Customer.class);
        q.where(b.like(q.from(Customer.class).<String> get("lastName"), b.literal("Bauer%"), '\\'));

        log.info("Test1 -- Customer found with criteria");
        em.createQuery(q).getResultList().forEach(bauer -> {
            log.info(bauer.toString());
        });

        log.info("Test2 -- Customer found with criteria");
        em.createQuery(q).getResultList().forEach(bauer -> {
            log.info(bauer.toString());
        });
    }
```

Versions information:

- hibernate-core: 5.6.7.Final (tested also with 5.6.6.Final)
- spring-data 2.6.2 / 2.6.3
- java 1.8.0_322 (temurin / adoptium)
- windows / linux (wsl)

[1]: (https://github.com/spring-projects/spring-data-jpa/files/8325311/demo_spring-data-2.6.3_hibernate_5.6.7.zip)

**Comment by aspan** [comment]
I have the same problem and i have debugged with different versions of hibernate. It works fine in hibernate 5.6.5 but fails in 5.6.6 or later.

Here is a minimal reproducer project:

[demo.zip](https://github.com/spring-projects/spring-data-jpa/files/8341124/demo.zip)
.

**Comment by knoobie** [comment]
We faced the same problem updating spring boot to 2.6.5 where the updated version of hibernate is used by default. 

Additionally we observed the problem with contains, startsWith and EqualsIgnoreCase as well.

**Comment by schauder** [comment]
This seems to be a Hibernate problem. 
I created an issue: https://hibernate.atlassian.net/browse/HHH-15142
And a reproducer: https://github.com/schauder/issue-HHH-15142 based on the reproducer of @aspan

**Comment by ettavolt** [comment]
I'd greatly appreciate if someone could help me promote the fix: hibernate/hibernate-orm#4918.

**Comment by fistons** [comment]
A workaround would be to explicitly add `@Param()` for each argument (see https://github.com/spring-projects/spring-data-jpa/issues/2479)

**Comment by THD-Thomas-Lang** [comment]
Another workaround seems to set the hibernate version (in Spring Boot´s pom.xml file) back to a working one:
`<hibernate.version>5.6.5.Final</hibernate.version>`

**Comment by 4braincells** [comment]
workaround is fine ... but why depending on a buggy hibernate version when everyone updates spring boot to 2.6.6 because of vulnerabilty issues ... <:o(

**Comment by C-Otto** [comment]
This is fixed in Hibernate 5.6.9 (according to https://hibernate.atlassian.net/browse/HHH-15142)

**Comment by basharatali-10p** [comment]
I got same issue but i resolve it by implementing the JPQL and did not use the 'startwith' function. Below is the example query
"Select c from Case c where  c.name like :name%"

**Comment by basharataliv** [comment]
I got same issue but i resolve it by implementing the JPQL and did not use the 'startwith' function. Below is the example query
"Select c from Case c where c.name like :name%"

**Comment by fistons** [comment]
Still have the issue with Spring Boot 2.6.7

**Comment by C-Otto** [comment]
> Still have the issue with Spring Boot 2.6.7

Spring Boot 2.6.7 contains Hibernate 5.6.8, and this bug is fixed in Hibernate 5.6.9.

**Comment by fistons** [comment]
> Spring Boot 2.6.7 contains Hibernate 5.6.8, and this bug is fixed in Hibernate 5.6.9.
Ok thanks! I guess I will have to wait for the next Spring Boot release

**Comment by moonfruit** [comment]
Not release yet.

**Comment by husniddinprogrammer** [comment]
🎉🎉🎉🎉🎉🧑‍💻I found the answer but it doesn't depend on the hibernate.Spring boot doesn't work on bigger versions 2.6.So change Spring boot to version 2.5.4.So it works 100 %.🧑‍💻

**Comment by husniddinprogrammer** [comment]
https://github.com/spring-projects/spring-data-jpa/issues/2472#issuecomment-1112918080

**Comment by husniddinprogrammer** [comment]
https://github.com/spring-projects/spring-data-jpa/issues/2472#issuecomment-1112918080

**Comment by 4braincells** [comment]
With gradle, just do this:

`implementation 'org.hibernate:hibernate-core:5.6.5.Final'

Should work with any spring (boot) version.
I think this had been said before...

**Comment by pankaj046** [comment]
> Has hibernate 5.6.9 released yet?. I cannot find it in the main maven repo (https://mvnrepository.com/artifact/org.hibernate/hibernate-core), the latest version is 5.6.8

use spring boot 2.6.2 version

**Comment by gdisanto** [comment]
> > Has hibernate 5.6.9 released yet?. I cannot find it in the main maven repo (https://mvnrepository.com/artifact/org.hibernate/hibernate-core), the latest version is 5.6.8
> 
> use spring boot 2.6.2 version

The project I am working is not using Spring boot, plus I do not want to downgrade my version of hibernate core.

**Comment by gregturn** [comment]
Workaround is to either pin Hibernate to an older, working version:

```xml
<properties>
    <hibernate.version>5.6.5.Final</hibernate.version>
</properties>
```
...or if you don't want to go to an older version of Hibernate, to parameterize your repository...

```java
@Repository
public interface BookRepository extends JpaRepository<Book, String> {

	List<Book> findByIsbnContaining(@Param("isbn") String isbn);
}
```
This forces query derivation to plugin the "right" parameter and hence not break at the Entity Manager.

**Comment by gregturn** [comment]
Follow #2519.

**Comment by Muchomor** [comment]
5.6.9 version does not seem to resolve this problem...

**Comment by fistons** [comment]
Sprint Boot 2.6.8 fixes the issue!

**Comment by mmm8955405** [comment]
> I'd greatly appreciate if someone could help me promote the fix: [hibernate/hibernate-orm#4918](https://github.com/hibernate/hibernate-orm/pull/4918).

The hibernate-orm tools cannot report problems in github. They are inefficient!

They generate documentation and examples of DAOs. At present, it can only run normally under the ant plug-in and the corresponding life cycle<goal>run</global>. The hibernate tools maven cannot configure the DAO location, which is very difficult to use. And they are changing their name (jakarta). God. A bunch of bugs

**Comment by abhisheksangani** [comment]
create a repository method using the @Query annotation for your signature
